src/coreclr/jit/llvmlssa.cpp

```diff
@@ -761,6 +767,62 @@ void Llvm::displayInitKindForLocal(unsigned lclNum, ValueInitKind initKind)
 }
 #endif // DEBUG
 
+//------------------------------------------------------------------------
+// getShadowFrameSize: What is the size of a function's shadow frame?
+//
+// Arguments:
+//    funcIdx - Index representing the function
+//
+// Return Value:
+//    The size of the shadow frame for the given function. We term this
+//    the value by which the shadow stack pointer must be offset before
+//    calling managed code such that the caller will not clobber anything
+//    live on the frame. Note that filters do not have any shadow state
+//    of their own and use the "original" frame from the parent function.
+//
+unsigned Llvm::getShadowFrameSize(unsigned funcIdx) const
+{
+    if (_compiler->funGetFunc(funcIdx)->funKind == FUNC_FILTER)
+    {
+        return 0;
+    }
+
+    assert((_shadowStackLocalsSize % TARGET_POINTER_SIZE) == 0);
+    return _shadowStackLocalsSize;
+}
+
+unsigned Llvm::getCalleeShadowStackOffset(unsigned funcIdx, bool isTailCall) const
+{
+    if (isTailCall)
+    {
+        return 0;
+    }
+
+    return getShadowFrameSize(funcIdx);
+}
+
+//------------------------------------------------------------------------
+// canEmitCallAsShadowTailCall: Can a call be made into a shadow tail call?
+//
+// Arguments:
+//    callIsInTry    - Is the call in a protected region
+//    callIsInFilter - Is the call in a filter
+//
+// Return Value:
+//    Whether a call can be made without preserving the current shadow frame.
+//
+bool Llvm::canEmitCallAsShadowTailCall(bool callIsInTry, bool callIsInFilter) const
+{
+    // Address-exposed shadow state may be observed by the callee or filters that run in the first pass of EH.
+    if (m_anyAddressExposedShadowLocals)
+    {
+        return false;
+    }
+
+    // Both protected regions and filters induce exceptional flow that may return back to this method.
+    return !callIsInTry && !callIsInFilter;
```

Comment:
Technically, we need not care about filters here since they "don't have shadow state of their own", but it seems decidedly better to be explicit about this edge case.